### PR TITLE
De-duplicate Module.xml jar entries based on scope.

### PR DESCRIPTION
**Pull request by benmusson**
The logic:
1. Group the manifests by `jarName`.
2. For each `jarName`, find all the scopes where it is applied and append them together in a list using `flatMap`.
3. Convert this list of scopes to a set in order to deduplicate, then join back together into a string.

At this point the de-duplication is complete and the original issue is resolved. https://github.com/inductiveautomation/ignition-module-tools/issues/52

I think it would also be a good idea to force a sort order for the jars, so I added it in.

1. Sort by `scope` length (jars used in more places are listed first).
2. Sort by `scope` alphabetically
3. Sort by `jarName` alphabetically

---
Before:
```xml
<jar scope="C">core-client-0.1.0.jar</jar>
<jar scope="C">kotlin-stdlib-2.0.0.jar</jar>
<jar scope="C">annotations-13.0.jar</jar>
<jar scope="C">if97-steam-tables-client-0.1.0.jar</jar>
<jar scope="CDG">core-common-0.3.0.jar</jar>
<jar scope="CDG">if97-2.0.0.jar</jar>
<jar scope="CDG">kotlin-stdlib-2.0.0.jar</jar> 
<jar scope="CDG">annotations-13.0.jar</jar>
<jar scope="CDG">if97-steam-tables-common-0.1.0.jar</jar>
<jar scope="D">core-designer-0.3.0.jar</jar>
<jar scope="D">kotlin-stdlib-2.0.0.jar</jar>
<jar scope="D">annotations-13.0.jar</jar>
<jar scope="D">if97-steam-tables-designer-0.1.0.jar</jar>
<jar scope="G">core-gateway-0.3.0.jar</jar>
<jar scope="G">kotlin-stdlib-2.0.0.jar</jar>
<jar scope="G">annotations-13.0.jar</jar>
<jar scope="G">if97-steam-tables-gateway-0.1.0.jar</jar>
```

---

After:
```xml
<jar scope="CDG">annotations-13.0.jar</jar>
<jar scope="CDG">core-common-0.3.0.jar</jar>
<jar scope="CDG">if97-2.0.0.jar</jar>
<jar scope="CDG">if97-steam-tables-common-0.1.0.jar</jar>
<jar scope="CDG">kotlin-stdlib-2.0.0.jar</jar>
<jar scope="C">core-client-0.1.0.jar</jar>
<jar scope="C">if97-steam-tables-client-0.1.0.jar</jar>
<jar scope="D">core-designer-0.3.0.jar</jar>
<jar scope="D">if97-steam-tables-designer-0.1.0.jar</jar>
<jar scope="G">core-gateway-0.3.0.jar</jar>
<jar scope="G">if97-steam-tables-gateway-0.1.0.jar</jar>
```

### CODE REVIEW

**Comment by brianeray** [comment]
Hi @benmusson , thanks for the PR. Can you add a unit test in `WriteModuleXmlTest.kt`?

File: gradle-module-plugin/src/functionalTest/kotlin/io/ia/sdk/gradle/modl/task/WriteModuleXmlTest.kt [hunk 1]
```diff
@@ -188,13 +217,26 @@ class WriteModuleXmlTest : BaseTest() {
...
         replacements: Map<String, String> = mapOf(),
+        dependencies: Map<String, String> = mapOf(),
         dumpBuildScript: Boolean = false,
     ): String {
         val projectDir = generateModule(
             tempFolder.newFolder(dirName),
             replacements,
         )
 
+        dependencies.forEach { (scopes, dependency) ->
```
Comment (benmusson) [inline gradle-module-plugin/src/functionalTest/kotlin/io/ia/sdk/gradle/modl/task/WriteModuleXmlTest.kt:228]: Added the ability to included additional dependencies in the `build.gradle` files of the generated scope subprojects.

**Comment by benmusson** [comment]
@brianeray rebased/squashed

**Comment by brianeray** [comment]
@benmusson can you take a look at my stab in #56? I'll probably pull in a reviewer from our side too, someone who knows module dev, because I'm a build engineer.

My focus is too narrow. I can do stuff in Gradle and JVM languages but not have the bigger picture in focus.

**Review by brianeray — APPROVED** [review]

**Pull request merged**